pkg/storage/prunable/prunable.go

```diff
@@ -118,3 +144,78 @@ func (p *Prunable) Flush() {
 		p.errorHandler(err)
 	}
 }
+
+func (p *Prunable) Rollback(targetSlotIndex iotago.SlotIndex) error {
+	timeProvider := p.apiProvider.APIForSlot(targetSlotIndex).TimeProvider()
+	targetSlotEpoch := timeProvider.EpochFromSlot(targetSlotIndex)
+	lastCommittedEpoch := targetSlotEpoch
+	// if the target index is the last slot of the epoch, the epoch was committed
+	if timeProvider.EpochEnd(targetSlotEpoch) != targetSlotIndex {
+		lastCommittedEpoch--
+	}
+
+	if err := p.prunableSlotStore.RollbackBucket(targetSlotEpoch, targetSlotIndex, timeProvider.EpochEnd(targetSlotEpoch)); err != nil {
+		return ierrors.Wrapf(err, "error while rolling back slots in a bucket for epoch %d", targetSlotEpoch)
+	}
+
+	// Shut down the prunableSlotStore in order to flush and get consistent state on disk after reopening.
+	p.prunableSlotStore.Shutdown()
+
+	// Removed entries that belong to the old fork and cannot be re-used.
+	for epochIdx := lastCommittedEpoch + 1; ; epochIdx++ {
+		fmt.Println("rollback before if", epochIdx, targetSlotEpoch)
+		if epochIdx > targetSlotEpoch {
+			shouldRollback, err := p.shouldRollbackCommittee(epochIdx, targetSlotIndex)
+			if err != nil {
+				return ierrors.Wrapf(err, "error while checking if committee for epoch %d should be rolled back", epochIdx)
+			}
+
+			fmt.Println("rollback committee", shouldRollback, "epochIdx", epochIdx, "lastCommittedEpoch", lastCommittedEpoch, "targetSlotEpoch", targetSlotEpoch)
+			if shouldRollback {
+				if err := p.committee.DeleteEpoch(epochIdx); err != nil {
+					return ierrors.Wrapf(err, "error while deleting committee for epoch %d", epochIdx)
+				}
+			}
+
+			if deleted := p.prunableSlotStore.DeleteBucket(epochIdx); !deleted {
+				break
+			}
+		}
+
+		if err := p.poolRewards.DeleteEpoch(epochIdx); err != nil {
+			return ierrors.Wrapf(err, "error while deleting pool rewards for epoch %d", epochIdx)
+		}
+		if err := p.poolStats.DeleteEpoch(epochIdx); err != nil {
+			return ierrors.Wrapf(err, "error while deleting pool stats for epoch %d", epochIdx)
+		}
+
+		if err := p.decidedUpgradeSignals.DeleteEpoch(epochIdx); err != nil {
+			return ierrors.Wrapf(err, "error while deleting decided upgrade signals for epoch %d", epochIdx)
+		}
+	}
+
+	return nil
+}
+
+// Remove committee for the next epoch only if forking point is before point of no return and committee is reused.
+// Always remove committees for epochs that are newer than targetSlotEpoch+1.
+func (p *Prunable) shouldRollbackCommittee(epochIndex iotago.EpochIndex, targetSlotIndex iotago.SlotIndex) (bool, error) {
```

Comment:
This logic is very specific and it might lead to issues if the behavior of the committee rotation changes. Maybe we can put it in the `Sybilprotection` and hand it in as a callback?

Reply:
I think we should be able to put it there: `PerformanceTracker` has the committee store already: https://github.com/iotaledger/iota-core/blob/3a3b48e0e53a0d002f660d1214e8b2ec51a8982e/pkg/protocol/sybilprotection/sybilprotectionv1/performance/performance.go#L20

The only thing we'd need to do is expose this method on the `SybilProtection` interface and hand it to the prunable.